pkg/login/social/connectors/org_role_mapper.go

```diff
@@ -0,0 +1,185 @@
+package connectors
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const MapperMatchAllOrgID = -1
+
+type OrgRoleMapper struct {
+	cfg        *setting.Cfg
+	logger     log.Logger
+	orgService org.Service
+}
+
+func ProvideOrgRoleMapper(cfg *setting.Cfg, orgService org.Service) *OrgRoleMapper {
+	return &OrgRoleMapper{
+		cfg:        cfg,
+		logger:     log.New("orgrole.mapper"),
+		orgService: orgService,
+	}
+}
+
+func (m *OrgRoleMapper) MapOrgRoles(ctx context.Context, externalOrgs []string, orgMappingSettings []string, directlyMappedRole org.RoleType) (map[int64]org.RoleType, error) {
+	orgMapping := m.splitOrgMappingSettings(ctx, orgMappingSettings)
+
+	userOrgRoles := getMappedOrgRoles(externalOrgs, orgMapping)
+
+	m.handleGlobalOrgMapping(userOrgRoles)
+
+	if len(userOrgRoles) == 0 {
+		userOrgRoles = m.GetDefaultOrgMapping(directlyMappedRole)
+	}
+
+	if directlyMappedRole == "" {
+		m.logger.Debug("No direct role mapping found")
+		return userOrgRoles, nil
+	}
+
+	m.logger.Debug("Direct role mapping found", "role", directlyMappedRole)
+
+	// Merge roles from org mapping `org_mapping` with role from direct mapping
+	for orgID, role := range userOrgRoles {
+		userOrgRoles[orgID] = getTopRole(directlyMappedRole, role)
+	}
+
+	return userOrgRoles, nil
+}
+
+func (m *OrgRoleMapper) GetDefaultOrgMapping(directlyMappedRole org.RoleType) map[int64]org.RoleType {
+	orgRoles := make(map[int64]org.RoleType, 0)
+
+	orgID := int64(1)
+	if m.cfg.AutoAssignOrg && m.cfg.AutoAssignOrgId > 0 {
+		orgID = int64(m.cfg.AutoAssignOrgId)
+	}
+
+	if directlyMappedRole == "" || !directlyMappedRole.IsValid() {
+		orgRoles[orgID] = org.RoleType(m.cfg.AutoAssignOrgRole)
+	} else {
+		orgRoles[orgID] = directlyMappedRole
+	}
+
+	return orgRoles
+}
+
+func (m *OrgRoleMapper) handleGlobalOrgMapping(orgRoles map[int64]org.RoleType) {
+	// No global role mapping => return
+	globalRole, ok := orgRoles[MapperMatchAllOrgID]
+	if !ok {
+		return
+	}
+
+	allOrgIDs, err := m.getAllOrgs()
+	if err != nil {
+		// Prevent resetting assignments
+		orgRoles = map[int64]org.RoleType{}
+		m.logger.Warn("error fetching all orgs, removing org mapping to prevent org sync")
```

Comment:
I think we can return here

Reply:
What happens with the providers if you have matched with role_attribute_path, but you get an error with `getAllOrgs`, does it remove you from any organizations you were mapped to previously?

Reply:
> I think we can return here

Agree, done!

> What happens with the providers if you have matched with role_attribute_path, but you get an error with getAllOrgs, does it remove you from any organizations you were mapped to previously?

Yes, the result will be the matched role assigned to the default organization. Added a test for this scenario: [should skip map to all organizations if org service returns an error](https://github.com/grafana/grafana/pull/86973/files#diff-82aefe39ddd11d74924f304762366c91879949c20b6ff0a1eaa6cafd9c272c8fR138)

Reply:
> Yes, the result will be the matched role assigned to the default organization. Added a test for this scenario: [should skip map to all organizations if org service returns an error](https://github.com/grafana/grafana/pull/86973/files#diff-82aefe39ddd11d74924f304762366c91879949c20b6ff0a1eaa6cafd9c272c8fR138)

That's an edge case for sure, but it does look a bit weird to me. As a user that was previously mapped as Admin to org:3 I can now become Viewer in org:1 because of a database error while fetching the orgs. I'd personally prefer the org mapping to be unchanged.

If we stick to that behavior, the log message needs to be changed.

Reply:
I checked the SAML implementation again and it does prevent resetting the assignment, so changed the logic to leave the org mapping unchanged in case global org mapping fails (when `getAllOrgs` returns an error).